File: java/src/main/java/creational/AbstractFactory.java
package creational;

// === Products ===
interface Button {

    void render();
}

interface Checkbox {

    void render();
}

class WindowsButton implements Button {

    @Override
    public void render() {
        System.out.println("Win button");
    }
}

class WindowsCheckbox implements Checkbox {

    @Override
    public void render() {
        System.out.println("Win checkbox");
    }
}

class LinuxButton implements Button {

    @Override
    public void render() {
        System.out.println("Linux button");
    }
}

class LinuxCheckbox implements Checkbox {

    @Override
    public void render() {
        System.out.println("Linux checkbox");
    }
}

// === Abstract factory ===
interface GUIFactory {

    Button createButton();

    Checkbox createCheckbox();
}

// === Concrete factories ===
class WindowsGUIFactory implements GUIFactory {

    @Override
    public Button createButton() {
        return new WindowsButton();
    }

    @Override
    public Checkbox createCheckbox() {
        return new WindowsCheckbox();
    }
}

class LinuxGUIFactory implements GUIFactory {

    @Override
    public Button createButton() {
        return new LinuxButton();
    }

    @Override
    public Checkbox createCheckbox() {
        return new LinuxCheckbox();
    }
}

@SuppressWarnings("unused")
final class AbstractFactoryExample {

    private AbstractFactoryExample() {
    }

    public static void example() {
        GUIFactory factory = detectFactory(); // could come from CONFIG/ENV
        Button btn = factory.createButton();
        Checkbox chk = factory.createCheckbox();
        btn.render();
        chk.render();
    }

    private static GUIFactory detectFactory() {
        String os = System.getProperty("os.name").toLowerCase();
        if (os.contains("win")) {
            return new WindowsGUIFactory();
        } else {
            return new LinuxGUIFactory();
        }
    }
}

File: rust/src/creational/abstract_factory.rs
trait Button {
    fn render(&self);
}
trait Checkbox {
    fn render(&self);
}

struct WinButton;
impl Button for WinButton {
    fn render(&self) {
        println!("Win button");
    }
}
struct WinCheckbox;
impl Checkbox for WinCheckbox {
    fn render(&self) {
        println!("Win checkbox");
    }
}

struct LinuxButton;
impl Button for LinuxButton {
    fn render(&self) {
        println!("Linux button");
    }
}
struct LinuxCheckbox;
impl Checkbox for LinuxCheckbox {
    fn render(&self) {
        println!("Linux checkbox");
    }
}

// === Abstract factory ===
trait GUIFactory {
    fn create_button(&self) -> Box<dyn Button>;
    fn create_checkbox(&self) -> Box<dyn Checkbox>;
}

// === Concrete factories ===
struct WindowsFactory;
impl GUIFactory for WindowsFactory {
    fn create_button(&self) -> Box<dyn Button> {
        Box::new(WinButton)
    }
    fn create_checkbox(&self) -> Box<dyn Checkbox> {
        Box::new(WinCheckbox)
    }
}

struct LinuxFactory;
impl GUIFactory for LinuxFactory {
    fn create_button(&self) -> Box<dyn Button> {
        Box::new(LinuxButton)
    }
    fn create_checkbox(&self) -> Box<dyn Checkbox> {
        Box::new(LinuxCheckbox)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example() {
        let os = "linux"; // should come from config
        let factory: Box<dyn GUIFactory> = match os {
            "windows" => Box::new(WindowsFactory),
            _ => Box::new(LinuxFactory),
        };

        let btn = factory.create_button();
        let chk = factory.create_checkbox();
        btn.render();
        chk.render();
    }
}
